src/physics/initial_conditions/initial_condition_function.cpp

```diff
@@ -450,6 +450,173 @@ ::value(const dealii::Point<dim,real> &point, const unsigned int istate) const
+    return soln_conservative[istate];
+}
+
+
+// ========================================================
+// 2D Low Density Euler -- Initial Condition
+// See Zhang & Shu, On positivity-preserving..., 2010 Pg. 10
+// ========================================================
+template <int dim, int nstate, typename real>
+InitialConditionFunction_LowDensity2D<dim,nstate,real>
+::InitialConditionFunction_LowDensity2D(
+    Parameters::AllParameters const* const param)
+    : InitialConditionFunction<dim, nstate, real>()
+{
+    // Euler object; create using dynamic_pointer_cast and the create_Physics factory
+    // This test should only be used for Euler
+    this->euler_physics = std::dynamic_pointer_cast<Physics::Euler<dim, dim + 2, double>>(
+        Physics::PhysicsFactory<dim, dim + 2, double>::create_Physics(param));
+}
+
+template <int dim, int nstate, typename real>
+inline real InitialConditionFunction_LowDensity2D<dim,nstate,real>
+::value(const dealii::Point<dim,real> &point, const unsigned int istate) const
+{
+    real density = 1+0.99*sin(point[0]+point[1]);
+    real vel_x = 1.0;
+    real vel_y = 1.0;
+    real pressure = 1.0;
+
+    std::array<real, nstate> soln_primitive;
+    soln_primitive[0] = density;
+    soln_primitive[1] = vel_x;
+    soln_primitive[2] = vel_y;
+    soln_primitive[3] = pressure;
+
+    const std::array<real, nstate> soln_conservative = this->euler_physics->convert_primitive_to_conservative(soln_primitive);
+    return soln_conservative[istate];
+}
+
+// ========================================================
+// 1D Leblanc Shock tube -- Initial Condition
+// See Zhang & Shu, On positivity-preserving..., 2010 Pg. 14
+// ========================================================
+template <int dim, int nstate, typename real>
+InitialConditionFunction_LeblancShockTube<dim,nstate,real>
+::InitialConditionFunction_LeblancShockTube(
+    Parameters::AllParameters const* const param)
+    : InitialConditionFunction<dim, nstate, real>()
+{
+    // Euler object; create using dynamic_pointer_cast and the create_Physics factory
+    // This test should only be used for Euler
+    this->euler_physics = std::dynamic_pointer_cast<Physics::Euler<dim, dim + 2, double>>(
+        Physics::PhysicsFactory<dim, dim + 2, double>::create_Physics(param));
+}
+
+template <int dim, int nstate, typename real>
+inline real InitialConditionFunction_LeblancShockTube<dim,nstate,real>
+::value(const dealii::Point<dim,real> &point, const unsigned int istate) const
+{
+    real density = 0.0;
+    real vel = 0.0;
+    real pressure = 0.0;
+
+    if(point[0] < 0.0){
+        density = 2.0;
+        vel = 0.0;
+        pressure = pow(10.0,9.0);
+    }
+    else{
+        density = 0.001;
+        vel = 0.0;
+        pressure = 1;
+    }
+
+    std::array<real, nstate> soln_primitive;
+    soln_primitive[0] = density;
+    soln_primitive[1] = vel;
+    soln_primitive[2] = pressure;
+
```

Comment:
you could create a base class of type InitialConditionFunction_Euler which provides the definition for `value` but calls a pure virtual function `primitive_value` which these three classes all derive from and provide the definition; this way the repetition between the classes is minimized and makes it easier for someone else to add an euler-type IC from primitive values

Reply:
Created a base class of type `InitialConditionFunction_EulerBase` and modified Sod, Leblanc, Shu Osher and 2D Low Density case to be derived from the new base class.